// vue3-hash-calendar/lib/CalendarYearMonth.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stdin_exports = {};
__export(stdin_exports, {
  calendarYearMonthProps: () => calendarYearMonthProps,
  default: () => stdin_default
});
module.exports = __toCommonJS(stdin_exports);
var import_vue = require("vue");
var import_vue2 = require("vue");
var import_CalendarScrollContainer = __toESM(require("./CalendarScrollContainer"));
var import_language = __toESM(require("./language"));
var import_utils = require("./utils");
const calendarYearMonthProps = {
  minDate: (0, import_utils.makeDateProp)(null),
  maxDate: (0, import_utils.makeDateProp)(null),
  disabledDate: {
    type: Function,
    default: () => false
  },
  checkedDayClassName: (0, import_utils.makeStringProp)(""),
  notCurrentMonthDayClassName: (0, import_utils.makeStringProp)(""),
  disabledClassName: (0, import_utils.makeStringProp)(""),
  type: (0, import_utils.makeStringProp)("date"),
  calendarContentHeight: (0, import_utils.makeNumberProp)(0),
  disabledScroll: {
    type: [Boolean, String],
    default: () => false
  },
  currentYearMonth: {
    type: Object,
    default: () => ({})
  },
  lang: (0, import_utils.makeStringProp)("CN")
};
var stdin_default = (0, import_vue2.defineComponent)({
  name: "CalendarYearMonth",
  props: calendarYearMonthProps,
  emits: ["click", "slidechange", "touchstart", "touchmove", "touchend"],
  setup(props, {
    emit
  }) {
    const language = (0, import_vue2.ref)({});
    const yearRange = (0, import_vue2.ref)(10);
    const disabledScrollDirec = (0, import_vue2.ref)(false);
    const yearMonthShow = (0, import_vue2.ref)([]);
    const lang = props.lang.toUpperCase();
    language.value = import_language.default[lang];
    const itemHeight = (0, import_vue2.computed)(() => props.calendarContentHeight / 4);
    const initYear = (year) => {
      const yearArr = [];
      const currYear = `${year || props.currentYearMonth.year}`;
      const yearStart = parseInt(currYear.substring(0, 3) + "0", 10);
      for (let i = 0; i <= yearRange.value; i++) {
        yearArr.push(yearStart + i);
      }
      yearArr.unshift(yearStart - 1);
      return yearArr;
    };
    const initYearRange = (year) => {
      const yearRangeArr = [];
      const currYear = `${year || props.currentYearMonth.year}`;
      const yearStart = parseInt(currYear.substring(0, 2) + "00", 10);
      for (let i = 0; i <= yearRange.value; i++) {
        yearRangeArr.push({
          s: yearStart + i * 10,
          e: yearStart + i * 10 + 9
        });
      }
      yearRangeArr.unshift({
        s: yearStart - 10,
        e: yearStart - 1
      });
      return yearRangeArr;
    };
    const getThreeYearArr = (year = props.currentYearMonth.year) => {
      const yearStr = year + "";
      const yearStart = yearStr.substring(0, 3);
      const yearStartLast = parseInt(parseInt(yearStart, 10) - 1 + "0", 10);
      const yearStartCurr = parseInt(yearStart + "0", 10);
      const yearStartNext = parseInt(parseInt(yearStart, 10) + 1 + "0", 10);
      return [initYear(yearStartLast), initYear(yearStartCurr), initYear(yearStartNext)];
    };
    const getThreeYearRangeArr = (year = props.currentYearMonth.year) => {
      const yearStr = year + "";
      const yearStart = yearStr.substring(0, 2);
      const yearStartLast = parseInt(parseInt(yearStart, 10) - 1 + "00", 10);
      const yearStartCurr = parseInt(yearStart + "00", 10);
      const yearStartNext = parseInt(parseInt(yearStart, 10) + 1 + "00", 10);
      return [initYearRange(yearStartLast), initYearRange(yearStartCurr), initYearRange(yearStartNext)];
    };
    const getNextOpitonData = () => {
      if (props.type === "year") {
        const year = yearMonthShow.value[2][1];
        yearMonthShow.value = getThreeYearArr(year);
      } else if (props.type === "yearRange") {
        const value = yearMonthShow.value[2][1];
        const year = value.s;
        yearMonthShow.value = getThreeYearRangeArr(year);
      }
    };
    const getLastOptionData = () => {
      if (props.type === "year") {
        const year = yearMonthShow.value[0][1];
        yearMonthShow.value = getThreeYearArr(year);
      } else if (props.type === "yearRange") {
        const value = yearMonthShow.value[0][1];
        const year = value.s;
        yearMonthShow.value = getThreeYearRangeArr(year);
      }
    };
    const slideChange = (direc) => {
      if (direc === "left") {
        getNextOpitonData();
      } else if (direc === "right") {
        getLastOptionData();
      }
      emit("slidechange", direc);
    };
    const getRangeYear = (date) => {
      const yearStart = date.s;
      const yearEnd = date.e;
      const yearArr = [];
      for (let i = yearStart; i <= yearEnd; i++) {
        yearArr.push(i);
      }
      return yearArr;
    };
    const isDisabled = (date, index) => {
      let fDate = new Date();
      if (props.type === "month") {
        fDate = new Date(`${props.currentYearMonth.year}/${index + 1}/01`);
      } else if (props.type === "year") {
        fDate = new Date(`${date}/${props.currentYearMonth.month + 1}/01`);
      } else if (props.type === "yearRange") {
        const yearArr = getRangeYear(date);
        return yearArr.every((year) => {
          fDate = new Date(`${year}/${props.currentYearMonth.month + 1}/01`);
          return props.disabledDate(fDate) || !(0, import_utils.isDateInRange)(fDate, props.minDate, props.maxDate);
        });
      }
      return props.disabledDate(fDate) || !(0, import_utils.isDateInRange)(fDate, props.minDate, props.maxDate);
    };
    const dateClick = (date, index) => {
      if (!date)
        return;
      if (isDisabled(date, index))
        return;
      let checkedDate = __spreadValues({}, props.currentYearMonth);
      if (props.type === "month") {
        checkedDate = __spreadProps(__spreadValues({}, checkedDate), {
          month: index
        });
      }
      if (props.type === "year") {
        checkedDate = __spreadProps(__spreadValues({}, checkedDate), {
          year: date
        });
      }
      if (props.type === "yearRange") {
        const yearArr = getRangeYear(date);
        checkedDate = __spreadProps(__spreadValues({}, checkedDate), {
          year: yearArr.includes(checkedDate.year) ? checkedDate.year : date.s
        });
      }
      emit("click", checkedDate, props.type);
    };
    const isChecked = (date, index) => {
      if (props.type === "month") {
        return index === props.currentYearMonth.month;
      }
      if (props.type === "year") {
        return date === props.currentYearMonth.year;
      }
      if (props.type === "yearRange") {
        return date.s <= props.currentYearMonth.year && date.e >= props.currentYearMonth.year;
      }
    };
    const isNotCurrent = (index) => (index === 0 || index === 11) && (props.type === "year" || props.type === "yearRange");
    const touchStart = (event) => {
      emit("touchstart", event);
    };
    const touchMove = (event) => {
      emit("touchmove", event);
    };
    const touchEnd = (event) => {
      emit("touchend", event);
    };
    (0, import_vue2.watch)(() => props.type, (val) => {
      disabledScrollDirec.value = props.disabledScroll;
      if (val === "month" || val === "date") {
        disabledScrollDirec.value = true;
        yearMonthShow.value = [language.value.MONTH, language.value.MONTH, language.value.MONTH];
      } else if (val === "year") {
        yearMonthShow.value = getThreeYearArr();
      } else if (val === "yearRange") {
        yearMonthShow.value = getThreeYearRangeArr();
      }
    });
    const renderYearMonthItem = (arr) => arr.map((item, index) => (0, import_vue.createVNode)("div", {
      "class": `year-body-item ${isDisabled(item, index) ? props.disabledClassName || "is_disabled" : ""}`,
      "style": {
        height: itemHeight.value + "px"
      },
      "key": index,
      "onClick": () => dateClick(item, index)
    }, [(0, import_vue.createVNode)("p", {
      "class": `year-body-item-content ${isChecked(item, index) ? props.checkedDayClassName || "is_checked" : ""} ${isNotCurrent(index) ? props.notCurrentMonthDayClassName || "is_not_current" : ""}`,
      "style": {
        width: props.type === "yearRange" ? "92px" : "60px"
      }
    }, [props.type === "yearRange" ? `${item.s}-${item.e}` : props.type === "month" ? language.value.MONTH[index] : item])]));
    const renderYearMonth = () => (0, import_vue.createVNode)("div", {
      "class": "year-body",
      "style": {
        height: itemHeight.value * 4 + "px",
        display: ["year", "yearRange", "month"].includes(props.type) ? "block" : "none"
      }
    }, [(0, import_vue.createVNode)(import_CalendarScrollContainer.default, {
      "calendarData": yearMonthShow.value,
      "disabledScroll": disabledScrollDirec.value,
      "onTouchstart": touchStart,
      "onTouchmove": touchMove,
      "onTouchend": touchEnd,
      "onSlidechange": slideChange
    }, (data) => renderYearMonthItem(data))]);
    return () => renderYearMonth();
  }
});
